x-pack/auditbeat/module/system/user/user.go

```diff
@@ -0,0 +1,303 @@
+	eventActionUserChanged = "user_changed"
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet collects data about a system's users.
+type MetricSet struct {
+	mb.BaseMetricSet
+	log    *logp.Logger
+	cache  *cache.Cache
+	bucket datastore.Bucket
+}
+
+// User represents a user. Fields according to getpwent(3).
+type User struct {
+	Name     string
+	Passwd   string
+	UID      uint32
+	GID      uint32
+	UserInfo string
+	Dir      string
+	Shell    string
+}
+
+// Hash creates a hash for User.
+func (user User) Hash() uint64 {
+	h := xxhash.New64()
+	// Use everything except userInfo
+	h.WriteString(user.Name)
+	h.WriteString(user.Passwd)
+	h.WriteString(strconv.Itoa(int(user.UID)))
+	h.WriteString(strconv.Itoa(int(user.GID)))
+	h.WriteString(user.Dir)
+	h.WriteString(user.Shell)
+	return h.Sum64()
+}
+
+func (user User) toMapStr() common.MapStr {
+	evt := common.MapStr{
+		"name":   user.Name,
+		"passwd": user.Passwd,
+		"uid":    user.UID,
+		"gid":    user.GID,
+		"dir":    user.Dir,
+		"shell":  user.Shell,
+	}
+
+	if user.UserInfo != "" {
+		evt.Put("user_information", user.UserInfo)
+	}
+
+	return evt
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		log:           logp.NewLogger(metricsetName),
+		cache:         cache.New(),
+		bucket:        bucket,
+	}
+
+	users, err := ms.restoreUsersFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore users from disk")
+	}
+	ms.log.Debugf("Restored %d users from disk", len(users))
+
+	ms.log.Debugf("%v", users)
+	ms.cache.DiffAndUpdateCache(convertToCacheable(users))
+
+	return ms, nil
+}
+
+// Load user cache from disk.
+func (ms *MetricSet) restoreUsersFromDisk() (users []*User, err error) {
+	var decoder *gob.Decoder
+	err = ms.bucket.Load("users", func(blob []byte) error {
+		if len(blob) > 0 {
+			buf := bytes.NewBuffer(blob)
+			decoder = gob.NewDecoder(buf)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if decoder != nil {
+		for {
+			user := new(User)
+			err = decoder.Decode(user)
+			if err == nil {
+				users = append(users, user)
+			} else if err == io.EOF {
+				// Read all users
+				break
+			} else {
+				return nil, errors.Wrap(err, "decode error")
+			}
+
+		}
+	}
+
+	return users, nil
+}
+
+// Save user cache to disk.
+func (ms *MetricSet) saveUsersToDisk(users []*User) error {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+
+	for _, user := range users {
+		err := encoder.Encode(*user)
+		if err != nil {
+			return errors.Wrap(err, "encode error")
+		}
+	}
+
+	ms.bucket.Store("users", buf.Bytes())
```

Comment:
This returns an error that is unhandled. Try using [gometalinter](https://github.com/alecthomas/gometalinter). It will run `errcheck` for you which points these kind of things out.

Reply:
Thanks - `gometalinter` is a handy tool. It's a shame that `errcheck` finds so many problems in the Beats codebase. Most commonly, we don't check the error returns of calls to `common.MapStr.Put` and `XXHash64.Write*`. Do you think we should be checking them, or `/nolint`-ing?

Reply:
IIRC XXHash64.Write does not ever return an error, but it has that signature to satisfy the `hash.Hash` interface. Likewise there are cases where MapStr.Put can never return an error like `Put("x", 1)` so skipping the check is fine.

If we were going to start linting all the time then yes figuring a way to squelch the false positives would be good. There's a tool called reviewdog (which `make lint`) uses that is supposed to only alert you to the linter warnings that were introduced on your branch. It compares the results from `master` against the current branch.